spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/RewritePositionDeleteFilesSparkAction.java

```diff
@@ -133,55 +133,59 @@ public RewritePositionDeleteFiles.Result execute() {
+    CloseableIterable<PositionDeletesScanTask> fileTasks = planFiles();
 
     try {
       StructType partitionType = Partitioning.partitionType(table);
-      StructLikeMap<List<PositionDeletesScanTask>> filesByPartition =
-          StructLikeMap.create(partitionType);
-
-      for (PositionDeletesScanTask task : scanTasks) {
-        StructLike coerced = coercePartition(task, partitionType);
-
-        List<PositionDeletesScanTask> partitionTasks = filesByPartition.get(coerced);
-        if (partitionTasks == null) {
-          partitionTasks = Lists.newArrayList();
-        }
-        partitionTasks.add(task);
-        filesByPartition.put(coerced, partitionTasks);
-      }
+      StructLikeMap<List<PositionDeletesScanTask>> fileTasksByPartition =
+          filesByPartition(partitionType, fileTasks);
 
-      StructLikeMap<List<List<PositionDeletesScanTask>>> fileGroupsByPartition =
-          StructLikeMap.create(partitionType);
-
-      filesByPartition.forEach(
-          (partition, partitionTasks) -> {
-            Iterable<List<PositionDeletesScanTask>> plannedFileGroups =
-                rewriter.planFileGroups(partitionTasks);
-            List<List<PositionDeletesScanTask>> groups = ImmutableList.copyOf(plannedFileGroups);
-            if (groups.size() > 0) {
-              fileGroupsByPartition.put(partition, groups);
-            }
-          });
-
-      return fileGroupsByPartition;
+      return fileGroupsByPartition(fileTasksByPartition);
     } finally {
       try {
-        scanTasks.close();
+        fileTasks.close();
       } catch (IOException io) {
         LOG.error("Cannot properly close file iterable while planning for rewrite", io);
       }
     }
   }
 
-  @VisibleForTesting
-  RewritePositionDeletesGroup rewriteDeleteFiles(
+  private CloseableIterable<PositionDeletesScanTask> planFiles() {
+    Table deletesTable =
+        MetadataTableUtils.createMetadataTableInstance(table, MetadataTableType.POSITION_DELETES);
+
+    return CloseableIterable.transform(
+        deletesTable.newBatchScan().ignoreResiduals().planFiles(),
+        t -> (PositionDeletesScanTask) t);
+  }
+
+  private StructLikeMap<List<PositionDeletesScanTask>> filesByPartition(
+      StructType partitionType, Iterable<PositionDeletesScanTask> tasks) {
+    StructLikeMap<List<PositionDeletesScanTask>> filesByPartition =
+        StructLikeMap.create(partitionType);
+
+    for (PositionDeletesScanTask task : tasks) {
+      StructLike coerced = coercePartition(task, partitionType);
+
+      List<PositionDeletesScanTask> partitionTasks = filesByPartition.get(coerced);
+      if (partitionTasks == null) {
+        partitionTasks = Lists.newArrayList();
+      }
+      partitionTasks.add(task);
+      filesByPartition.put(coerced, partitionTasks);
+    }
+
+    return filesByPartition;
+  }
+
+  private StructLikeMap<List<List<PositionDeletesScanTask>>> fileGroupsByPartition(
```

Comment:
What if we add `planFileGroups` like this?

```
private List<List<PositionDeletesScanTask>> planFileGroups(List<PositionDeletesScanTask> tasks) {
  return ImmutableList.copyOf(rewriter.planFileGroups(tasks));
}
```

Then we can call it directly:

```
return StructLikeMap.transformValues(fileTasksByPartition, this::planFileGroups);
```

All would fit on one line.